# Encoder Example

In [1]:
import torch
from transformers import AutoTokenizer, AutoModel

In [2]:
tokenizer = AutoTokenizer.from_pretrained("bert-base-uncased")
encoder_model = AutoModel.from_pretrained("bert-base-uncased")

e:\Important\Skills\llm-learn\env\Lib\site-packages\keras\src\export\tf2onnx_lib.py:8: FutureWarning: In the future `np.object` will be defined as the corresponding NumPy scalar.
  if not hasattr(np, "object"):


In [3]:
sentence = "The bank approved the loan"

inputs = tokenizer(sentence, return_tensors="pt")
outputs = encoder_model(**inputs)

In [4]:
print(outputs.last_hidden_state.shape)

torch.Size([1, 7, 768])


### Meaning

- **1 sentence**
- **7 tokens** (after subword tokenization)
- Each token → a **768-dimensional, context-aware embedding**

### What the Encoder Did

- Read the **entire sentence at once**
- Applied **self-attention**
- Created **context-aware representations**
- **Did not generate text**


# Decoder Example

In [5]:
from transformers import AutoModelForCausalLM

tokenizer = AutoTokenizer.from_pretrained("gpt2")
decoder_model = AutoModelForCausalLM.from_pretrained("gpt2")

Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`


model.safetensors:  11%|#1        | 62.9M/548M [00:00<?, ?B/s]

e:\Important\Skills\llm-learn\env\Lib\site-packages\huggingface_hub\file_download.py:143: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\Arshad Ziban\.cache\huggingface\hub\models--gpt2. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)


generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

In [8]:
prompt = "The bank approved"

inputs = tokenizer(prompt, return_tensors="pt")

In [9]:
output_ids = decoder_model.generate(
    **inputs,
    max_length=15,
    do_sample=False
)

generated_text = tokenizer.decode(output_ids[0], skip_special_tokens=True)
print(generated_text)

Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


The bank approved the loan in May, but the bank has not yet approved
